src/neo/Ledger/MemoryPool.cs

```diff
@@ -258,21 +259,28 @@ internal bool CanTransactionFitInPool(Transaction tx)
         /// Note: This must only be called from a single thread (the Blockchain actor). To add a transaction to the pool
         ///       tell the Blockchain actor about the transaction.
         /// </summary>
+        /// <param name="snapshot">Snapshot</param>
         /// <param name="hash"></param>
         /// <param name="tx"></param>
         /// <returns></returns>
-        internal bool TryAdd(UInt256 hash, Transaction tx)
+        internal bool TryAdd(SnapshotView snapshot, UInt256 hash, Transaction tx)
         {
             var poolItem = new PoolItem(tx);
 
+            _txRwLock.EnterWriteLock();
             if (_unsortedTransactions.ContainsKey(hash)) return false;
 
             List<Transaction> removedTransactions = null;
-            _txRwLock.EnterWriteLock();
             try
             {
+                // Check balance
+                BigInteger balance = NativeContract.GAS.BalanceOf(snapshot, tx.Sender);
                 _unsortedTransactions.Add(hash, poolItem);
-                SendersFeeMonitor.AddSenderFee(tx);
+                if (balance < SendersFeeMonitor.AddSenderFee(tx))
+                {
+                    _unsortedTransactions.Remove(hash);
+                    return false;
+                }
```

Comment:
I think this isn't the best way. And it's not enough when `balance < SendersFeeMonitor.AddSenderFee(tx)`, more txs will be removed from memory pool.

Reply:
Since we already have `UpdatePoolForBlockPersisted`, and this `TryAdd` is called from a single thread (the Blockchain actor).  We need to find the cause of this phenomenon.

https://github.com/neo-project/neo/blob/fc1a64b77c8a901c048e7b973f204d3b7c9870e3/src/neo/Ledger/MemoryPool.cs#L351-L361

Reply:
This happend because it's not a single thread, Blockchain actor, but was send by RPC 

https://github.com/neo-project/neo-modules/blob/ef91c522f13ef30bc6308a2e343a4c3466bb3885/src/RpcServer/RpcServer.Wallet.cs#L288

So we can add a lock in `OnNewTransaction` and check if it's solved

Reply:
> So we can add a lock in OnNewTransaction and check if it's solved
It doesn't work in my tests..